README.md
[![CRAN version](http://www.r-pkg.org/badges/version/basetheme)](https://cran.r-project.org/package=basetheme)
[![dependencies](https://tinyverse.netlify.com/badge/basetheme)](https://CRAN.R-project.org/package=basetheme)
[![Monthly Downloads](https://cranlogs.r-pkg.org/badges/basetheme)](https://cranlogs.r-pkg.org/badges/basetheme)

# basetheme #

R package implementing a theming system for base plotting.

## Description ##

The `basetheme()` function provides a **persistent** way to set and unset R plotting settings.
All the user has to do is set the theme once and it will be consistently applied to all the subsequent plots even after the graphics device is closed.
The function only works with base graphics system, so if you are exclusively working with `ggplot2`, boy, did you install the wrong package.

## Usage ##

The persistent settings are controlled via the `basetheme()` function.
To an extent this function mimics the behaviour and arguments of `par()` and has multiple modes of operation:

```r
# Specifying the `par()` values directly:
basetheme(pch=19, mgp=c(2,.7,0), tck=-.01)

# Choosing a theme by name:
basetheme("clean")

# Using a mix of the above:
basetheme("minimal", bg="grey", pch=1)

# Specifying a list with theme values:
theme <- basetheme("clean")
theme$cex.main <- 2
basetheme(theme)

# Obtaining a list of parameters for the current theme:
basetheme()

# Removing the current theme:
basetheme(NULL)
```

In addition to `basetheme()` two additional functions are provided for colors manipulation within the plots: `num2col()` and `lab2col()`.
See documentation: `help(basetheme)`, `help(num2col)`, `help(lab2col)` and examples: `example(basetheme)`, `example(num2col)`, `example(lab2col)` for more details.

## Some Examples ##

A few examples using of how it all works together.

```r
basetheme("void")
boxplot(split(iris$Sepal.Width, iris$Species))
```

![](http://karolis.koncevicius.lt/data/basetheme/examples/ex_1a.png)

```r
basetheme("clean")
barplot(rivers, col=num2col(rivers))
```

![](http://karolis.koncevicius.lt/data/basetheme/examples/ex_2a.png)

```r
basetheme("brutal")
plot(hclust(dist(USArrests), "ward.D2"), hang=-1)
```

![](http://karolis.koncevicius.lt/data/basetheme/examples/ex_3a.png)

```r
basetheme("royal")
pairs(iris[-5], bg=lab2col(iris$Species), col=0)
```

![](http://karolis.koncevicius.lt/data/basetheme/examples/ex_4a.png)

```r
basetheme("deepblue")
pairs(iris[-5], bg=num2col(iris[,1]), col=0)
```

![](http://karolis.koncevicius.lt/data/basetheme/examples/ex_5a.png)

```r
x <- seq(-1.95, 1.95, length = 30)
y <- seq(-1.95, 1.95, length = 35)
z <- outer(x, y, function(a, b) a*b^2)

basetheme("dark")
persp(x, y, z, theta=-45)
```

![](http://karolis.koncevicius.lt/data/basetheme/examples/ex_6a.png)


## More Details ##

Additional details can be found in the Wiki pages:

1. [List of Available Themes](https://github.com/karoliskoncevicius/basetheme/wiki/Theme-List)
2. [Creating Custom Themes](https://github.com/karoliskoncevicius/basetheme/wiki/Creating-Themes)
3. [Installation Instructions](https://github.com/karoliskoncevicius/basetheme/wiki/Installation)
4. [Implementation Details](https://github.com/karoliskoncevicius/basetheme/wiki/Implementation-Details)

## See Also ##

CRAN:

1. [**prettyB**](https://github.com/jumpingrivers/prettyB)

GitHub:

1. [**styles**](https://github.com/leehazel/styles)



NEWS
Package: basetheme
==================

Version 0.1.3 [2023-04-07]

BUG FIXES

* Theme palette is now set immediately instead of waiting for the first plot.

IMPROVEMENTS

 * Missing color in 'num2col' and 'lab2col' was set to grey in all themes.


Version 0.1.2 [2019-10-17]

NEW FEATURES

 * Function 'num2col' for turning numbers into colors.
 * Function 'lab2col' for turning labels into colors.

IMPROVEMENTS

 * Implemented consistent color palettes for all themes.

BUG FIXES

 * Fixed a bug that ignored par(new=TRUE).


Version 0.1.1 [2019-07-17]

NEW FEATURES

 * Theme: deepblue.
 * Theme: royal.

IMPROVEMENTS

 * Removed axis ticks from the "minimal" theme.
 * Adjusted "void" theme to be black and white.


Version 0.1.0 [2019-06-14]

NEW FEATURES

 * Theme: dark.
 * Theme: ink.
 * Theme: brutal.
 * Theme: clean.
 * Theme: void.
 * Theme: minimal.
 * Theme: default.
 * Created.


R/utils.R
# function to obtain the current theme settings
current_theme <- function() {
  hook1 <- getHook("before.plot.new")
  hook2 <- getHook("plot.new")

  theme <- list()
  if(is.function(hook1$par))
    theme <- c(theme, hook1$par(FALSE))
  if(is.function(hook1$palette))
    theme <- c(theme, hook1$palette(FALSE))
  if(is.function(hook2$rect))
    theme <- c(theme, hook2$rect(FALSE))

  theme
}

# function to get theme parameters by name
get_theme <- function(theme=NULL) {
  themes <- list_themes()
  if (theme %in% names(themes)) {
    themes[[theme]]()
  } else {
    stop("no such theme: ", theme, ".\nAvailable themes: ",
         paste(names(list_themes()), collapse=", ")
    )
  }
}

# function to set par settings
setParFun <- function(plist) {
  parList <- plist
  parList[grep("^palette\\.?", names(parList))] <- NULL
  parList[grep("^rect\\.", names(parList))] <- NULL
  function(set=TRUE) {
    if(set) {
      parList$new <- graphics::par('new')
      do.call(graphics::par, parList)
    } else {
      parList
    }
  }
}

# function to set palette settings
setPalFun <- function(plist) {
  plist <- plist[grep("^palette\\.?", names(plist))]
  if(!is.null(plist$palette)) {
    grDevices::palette(value=plist$palette)
    function(set=TRUE) {
      if(set)
        grDevices::palette(value=plist$palette)
      else
        plist
    }
  } else {
    NULL
  }
}

# function to set rect settings
setRecFun <- function(plist) {
  plist <- plist[grep("^rect\\.", names(plist))]
  if(length(plist) > 0) {
    function(set=TRUE) {
      if(set) {
        rlist <- list(xleft   = graphics::par("usr")[1], ybottom=graphics::par("usr")[3],
                      xright  = graphics::par("usr")[2], ytop=graphics::par("usr")[4],
                      col     = ifelse(is.null(plist$rect.col), NA, plist$rect.col),
                      density = ifelse(is.null(plist$rect.density), numeric(), plist$rect.density),
                      angle   = ifelse(is.null(plist$rect.angle), 35, plist$rect.angle),
                      lwd     = ifelse(is.null(plist$rect.lwd), 1, plist$rect.lwd),
                      lty     = ifelse(is.null(plist$rect.lty), 1, plist$rect.lty),
                      border  = ifelse(is.null(plist$rect.border), NA, plist$rect.border)
                      )
        do.call(graphics::rect, rlist)
      } else {
        plist
      }
    }
  } else {
    NULL
  }
}



R/num2col.R
#' Numeric Vector to Color
#'
#' Returns interpolated colors for a specified numeric vector.
#'
#' This function interpolates a given set of colors to a numeric vector.
#' Main use case is in turning numbers into colors for plots,
#' especially when different ranges of \code{x} has to be colored differently.
#'
#' @param x numeric vector (factors are transformed to numeric)
#' @param ... colors to build palette from (defaults to jet colors)
#' @param NAcol color to be used for NA values (defaults to grey)
#' @param xrange original range of x values (defaults to \code{range(x)})
#'
#' @return a vector of colors for each element in x.
#'
#' @examples
#'   plot(mtcars$hp, mtcars$mpg, col=num2col(mtcars$mpg), pch=19, cex=2)
#'
#'   cols <- c("lightblue", "cornflowerblue", "orange", "red")
#'   plot(mtcars$hp, mtcars$mpg, col=num2col(mtcars$mpg, cols), pch=19, cex=2)
#'
#'   plot(mtcars$hp, mtcars$mpg, col=num2col(mtcars$mpg, "green", "red"), pch=19, cex=2)
#'
#' @author Karolis Koncevičius
#' @export
num2col <- function(x, ..., NAcol="#B9BBB6", xrange=range(x, na.rm=TRUE)) {
  if(is.factor(x)) x <- as.numeric(x)
  if(length(xrange) != 2) stop('"xrange" must have 2 elements: min and max')
  if(min(x, na.rm=TRUE) < xrange[1]) stop('x has values lower than "xrange[1]"')
  if(max(x, na.rm=TRUE) > xrange[2]) stop('x has values higher than "xrange[2]"')

  pal <- c(...)
  if(length(pal) < 1) {
    pal <- c("blue", "#007FFF", "cyan", "#7FFF7F", "yellow", "#FF7F00", "red")
  }

  vals <- x - xrange[1]
  vals <- vals / (xrange[2]-xrange[1])

  ramp <- colorRamp(pal)
  cols <- vals
  cols[] <- NAcol
  cols[!is.na(vals)] <- rgb(ramp(vals[!is.na(vals)]), maxColorValue=255)
  cols
}



R/coltools.R
#' Numbers to Colors
#'
#' Assigns colors to the provided vector of numbers.
#'
#' This function interpolates a given set of colors to a numeric vector.
#' Main use case is in turning numbers into colors for plots,
#' especially when different ranges of \code{x} have to be colored differently.
#'
#' Color of NA values and values outside of \code{ref} range can be set using \code{NAcol} argument.
#' Set this to NA to omit the display of such values.
#'
#' In case only a single color is provided - it is expanded by using tints and shades.
#'
#' When \code{x} is not specified - a function that generates colors based on
#' \code{pal} and \code{ref} is returned.
#'
#' @param x numeric vector (factors are transformed to numeric)
#' @param pal colors used to build the palette (defaults to colors set by theme)
#' @param ref reference for assigning colors (defaults to the range of \code{x})
#' @param NAcol color to be used for NA values (defaults to color set by theme)
#'
#' @return a vector of colors for each element in \code{x} or, when \code{x} is missing, a function.
#'
#' @examples
#'  # color numbers by y-axos
#'  plot(mtcars$hp, mtcars$mpg, col=num2col(mtcars$mpg), pch=19)
#'
#'  # color only a certain range
#'  plot(mtcars$hp, mtcars$mpg, col=num2col(mtcars$mpg, ref=c(20, 35)), pch=19)
#'
#'  # hide the out of range values
#'  plot(mtcars$hp, mtcars$mpg, col=num2col(mtcars$mpg, ref=c(20,35), NAcol=NA))
#'
#'  # iris example
#'  pairs(iris[,-5], col=num2col(iris$Sepal.Length))
#'
#'  # same butusing a prepared coloring function (for values in range 0-10)
#'  color <- num2col(ref=c(0,10))
#'  pairs(iris[,-5], col=color(iris$Sepal.Length))
#'
#' @seealso \code{lab2col}
#'
#' @author Karolis Koncevičius
#' @export
num2col <- function(x, pal, ref=range(x, na.rm=TRUE), NAcol) {
  if(missing(pal)) {
    pal <- basetheme()$palette.numbers
  }
  if(is.null(pal)) {
    pal <- c("blue", "#007FFF", "cyan", "#7FFF7F", "yellow", "#FF7F00", "red")
  }

  if(length(pal) == 1) {
    pal <- colshade(pal, c(-0.8, -0.5, -0.3, 0, 0.3, 0.5, 0.8))
  }

  if(missing(NAcol)) {
    NAcol <- basetheme()$palette.missing
  }
  if(is.null(NAcol)) {
    NAcol <- "#B9BBB6"
  }

  ramp <- grDevices::colorRamp(pal)
  min <- min(ref, na.rm=TRUE)
  max <- max(ref, na.rm=TRUE)

  fun <- function(x) {
    if(is.factor(x)) x <- as.numeric(x)
    x[x < min | x > max] <- NA
    x <- x - min
    x <- x / (max-min)

    cols <- x
    cols[] <- NAcol
    cols[!is.na(x)] <- grDevices::rgb(ramp(x[!is.na(x)]), maxColorValue=255)
    cols
  }

  if(missing(x)) {
    fun
  } else {
    fun(x)
  }
}

#' Labels to Colors
#'
#' Assigns colors to the provided vector of labels.
#'
#' This function assigns colors to each unique level of label vector \code{x}.
#' Mainly used for consistently assigning colors to a unique set of labels,
#' especially when the order of original labels might change (see examples).
#'
#' Color of NA values and values outside of specified reference can be set using
#' \code{NAcol} argument. Set this to NA to omit the display of such values.
#'
#' \code{ref} parameter can be specified to select the order of color assignment:
#' first element from \code{ref} will be assigned the first color from \code{pal},
#' second element - second color, and so on.
#'
#' The list of provided colors is expanded by first adding shades and then adding
#' tints. However if the number of groups exceeds the number of provided colors
#' by more than 3 times the colors will be repeated.
#'
#' When \code{x} is not specified - a function that generates colors based on
#' \code{pal} and \code{ref} is returned.
#'
#' @param x vector of labels (always transformed to character)
#' @param pal colors used to build the palette (defaults to colors set by theme)
#' @param ref reference for assigning colors (defaults to elements of \code{x})
#' @param NAcol color to be used for NA values (defaults to color set by theme)
#'
#' @return a vector of colors for each element in \code{x} or, when \code{x} is missing, a function.
#'
#' @examples
#'  # iris example
#'  pairs(iris[,1:4], col=lab2col(iris$Species))
#'
#'  # iris example with one group missing from reference
#'  pairs(iris[,1:4], col=lab2col(iris$Species, ref=c("setosa", "versicolor")))
#'
#'  # example of using a coloring function
#'  # "color" function below will consistently assign colors to values in "ref".
#'  color <- lab2col(ref=unique(chickwts$feed))
#'  par(mfrow=c(1,2))
#'  means <- tapply(chickwts$weight, chickwts$feed, mean)
#'  barplot(means, col=color(names(means)), las=2)
#'  means <- sample(means)
#'  barplot(means, col=color(names(means)), las=2)
#'
#' @seealso \code{num2col}
#'
#' @author Karolis Koncevičius
#' @export
lab2col <- function(x, pal, ref=x, NAcol) {
  if(missing(pal)) {
    pal <- basetheme()$palette.labels
  }
  if(is.null(pal)) {
    pal <- c("#5DA5DA", "#FAA43A", "#60BD68", "#F15854", "#B276B2", "#8D4B08", "#DECF3F", "#F17CB0", "#66E3D9", "#00FF7F")
  }

  if(missing(NAcol)) {
    NAcol <- basetheme()$palette.missing
  }
  if(is.null(NAcol)) {
    NAcol <- "#B9BBB6"
  }

  pal <- rep_len(pal, length(pal) * 3)
  pal <- colshade(pal, rep(c(0, 0.5, -0.5), each=length(pal)/3))

  vals <- unique(ref)
  if(length(vals) > length(pal))
    warning("number of unique values exceeds the number of colors in the palette by more than 3 times")
  pal  <- rep_len(pal, length(vals))

  fun <- function(x) {
    if(is.numeric(x)) x <- as.character(x)

    cols <- pal[match(x, vals)]
    cols[is.na(cols)] <- NAcol
    cols
  }

  if(missing(x)) {
    fun
  } else {
    fun(x)
  }
}


#' Add Shade to Color
#'
#' Adds a selected amount of shade or tint to a vector of colors.
#'
#' This function adds shades and tints to provided list of colors.
#' Shade or tint is decided depending on the sign of the \code{frac} argument:
#' positive values make colors darker and negative values lighten them.
#'
#' @param cols a vector of colors
#' @param frac a vector of shade fractions (between -1 and 1, defaults to 0)
#'
#' @return a vector of colors with added shades or tints.
#'
#' @author Karolis Koncevičius
colshade <- function(cols, frac=0) {
  if(any(frac < -1 | frac > 1)) stop('"frac" must be between -1 and 1')
  if(length(frac) <= length(cols)) {
    frac <- rep_len(frac, length(cols))
  } else {
    cols <- rep_len(cols, length(frac))
  }

  addTint <- function(col, tint) {
    grDevices::rgb(t(grDevices::col2rgb(col)) + tint*t((255-grDevices::col2rgb(col))), maxColorValue=255)
  }
  addShade <- function(col, shade) {
    grDevices::rgb(t(grDevices::col2rgb(col)) * (1-shade), maxColorValue=255)
  }

  cols[frac <= 0] <- addTint(cols[frac <= 0], -frac[frac <= 0])
  cols[frac > 0]  <- addShade(cols[frac > 0], frac[frac > 0])
  cols
}



R/themes.R
basetheme_default <- function() {
  palette <- c("black", "red", "green3", "blue", "cyan", "magenta", "yellow", "gray")
  palette_numbers <- c()
  palette_labels  <- c()
  palette_missing <- c()

  list(fg              = "black",
       adj             = 0.5,
       ann             = TRUE,
       bg              = "transparent",
       bty             = "o",
       cex             = 1,
       cex.axis        = 1,
       cex.lab         = 1,
       cex.main        = 1.2,
       cex.sub         = 1,
       col             = "black",
       col.axis        = "black",
       col.lab         = "black",
       col.main        = "black",
       col.sub         = "black",
       family          = "",
       font            = 1,
       font.axis       = 1,
       font.lab        = 1,
       font.main       = 2,
       font.sub        = 1,
       lab             = c(5, 5, 7),
       las             = 0,
       lend            = "round",
       ljoin           = "round",
       lmitre          = 10,
       lty             = "solid",
       lwd             = 1,
       mgp             = c(3, 1, 0),
       pch             = 1,
       tck             = NA,
       xaxs            = "r",
       xaxt            = "s",
       xpd             = FALSE,
       yaxs            = "r",
       yaxt            = "s",
       rect.col        = NA,
       rect.density    = numeric(),
       rect.angle      = 45,
       rect.lwd        = 1,
       rect.lty        = 1,
       rect.border     = NA,
       palette         = palette,
       palette.numbers = palette_numbers,
       palette.labels  = palette_labels,
       palette.missing = palette_missing
       )
}


basetheme_minimal <- function() {
  palette         <- c("#4D4D4D", "#F15854", "#60BD68", "#5DA5DA", "#66E3D9", "#B276B2", "#DECF3F", "#B4B4B4", "#FAA43A", "#8D4B08")
  palette_numbers <- c("#4F61A1", "#5DA5DA", "#66E3D9", "#7FFF7F", "#FFF25C", "#FAA43A", "#F15854")
  palette_labels  <- c("#5DA5DA", "#FAA43A", "#60BD68", "#F15854", "#B276B2", "#8D4B08", "#DECF3F", "#F17CB0", "#66E3D9", "#00FF7F")
  palette_missing <- "#B4B4B4"

  list(fg              = "black",
       adj             = 0.5,
       ann             = TRUE,
       bg              = "white",
       bty             = "n",
       cex             = 1,
       cex.axis        = 0.8,
       cex.lab         = 1,
       cex.main        = 1.2,
       cex.sub         = 1,
       col             = "black",
       col.axis        = "black",
       col.lab         = "black",
       col.main        = "black",
       col.sub         = "black",
       family          = "serif",
       font            = 1,
       font.axis       = 1,
       font.lab        = 1,
       font.main       = 2,
       font.sub        = 2,
       lab             = c(4,4,7),
       las             = 1,
       lend            = 0,
       ljoin           = 0,
       lmitre          = 10,
       lty             = 1,
       lwd             = 1,
       mgp             = c(2.5,0.2,0),
       pch             = 20,
       tck             = 0,
       xaxs            = "r",
       xaxt            = "s",
       xpd             = FALSE,
       yaxs            = "r",
       yaxt            = "s",
       rect.col        = NA,
       rect.density    = numeric(),
       rect.angle      = 45,
       rect.lwd        = 1,
       rect.lty        = 1,
       rect.border     = NA,
       palette         = palette,
       palette.numbers = palette_numbers,
       palette.labels  = palette_labels,
       palette.missing = palette_missing
  )
}


basetheme_void <- function() {
  palette         <- c("#4D4D4D", "#F15854", "#60BD68", "#5DA5DA", "#66E3D9", "#B276B2", "#DECF3F", "#B4B4B4", "#FAA43A", "#8D4B08")
  palette_numbers <- c("#4F61A1", "#5DA5DA", "#66E3D9", "#7FFF7F", "#FFF25C", "#FAA43A", "#F15854")
  palette_labels  <- c("#5DA5DA", "#FAA43A", "#60BD68", "#F15854", "#B276B2", "#8D4B08", "#DECF3F", "#F17CB0", "#66E3D9", "#00FF7F")
  palette_missing <- "#B4B4B4"

  list(fg              = "black",
       adj             = 0.5,
       ann             = FALSE,
       bg              = "white",
       bty             = "n",
       cex             = 1,
       cex.axis        = 1.1,
       cex.lab         = 1.1,
       cex.main        = 1.5,
       cex.sub         = 1,
       col             = "black",
       col.axis        = "black",
       col.lab         = "black",
       col.main        = "black",
       col.sub         = "black",
       family          = "",
       font            = 1,
       font.axis       = 1,
       font.lab        = 1,
       font.main       = 2,
       font.sub        = 2,
       lab             = c(5,5,7),
       las             = 1,
       lend            = 0,
       ljoin           = 0,
       lmitre          = 10,
       lty             = 1,
       lwd             = 2,
       mgp             = c(3,0.5,0),
       pch             = 19,
       tck             = 0,
       xaxs            = "r",
       xaxt            = "n",
       xpd             = FALSE,
       yaxs            = "r",
       yaxt            = "n",
       rect.col        = NA,
       rect.density    = numeric(),
       rect.angle      = 45,
       rect.lwd        = 1,
       rect.lty        = 1,
       rect.border     = NA,
       palette         = palette,
       palette.numbers = palette_numbers,
       palette.labels  = palette_labels,
       palette.missing = palette_missing
       )
}


basetheme_clean <- function() {
  palette         <- c("#4D4D4D", "#F15854", "#60BD68", "#5DA5DA", "#66E3D9", "#B276B2", "#DECF3F", "#B4B4B4", "#FAA43A", "#8D4B08")
  palette_numbers <- c("#4F61A1", "#5DA5DA", "#66E3D9", "#7FFF7F", "#FFF25C", "#FAA43A", "#F15854")
  palette_labels  <- c("#5DA5DA", "#FAA43A", "#60BD68", "#F15854", "#B276B2", "#8D4B08", "#DECF3F", "#F17CB0", "#66E3D9", "#00FF7F")
  palette_missing <- "#B4B4B4"

  list(fg              = "black",
       adj             = 0.5,
       ann             = TRUE,
       bg              = "white",
       bty             = "o",
       cex             = 1,
       cex.axis        = 0.7,
       cex.lab         = 0.8,
       cex.main        = 1,
       cex.sub         = 0.8,
       col             = "black",
       col.axis        = "black",
       col.lab         = "black",
       col.main        = "black",
       col.sub         = "black",
       family          = "",
       font            = 1,
       font.axis       = 1,
       font.lab        = 1,
       font.main       = 2,
       font.sub        = 3,
       lab             = c(5, 5, 7),
       las             = 1,
       lend            = "round",
       ljoin           = "round",
       lmitre          = 10,
       lty             = "solid",
       lwd             = 1,
       mgp             = c(2, 0.5, 0),
       pch             = 20,
       tck             = -0.01,
       xaxs            = "r",
       xaxt            = "s",
       xpd             = FALSE,
       yaxs            = "r",
       yaxt            = "s",
       rect.col        = NA,
       rect.density    = numeric(),
       rect.angle      = 45,
       rect.lwd        = 1,
       rect.lty        = 1,
       rect.border     = NA,
       palette         = palette,
       palette.numbers = palette_numbers,
       palette.labels  = palette_labels,
       palette.missing = palette_missing
       )
}


basetheme_brutal <- function() {
  palette         <- c("#303643", "#BF616A", "#8FA87A", "#5E81AC", "#66CDAA", "#D070B9", "#E7D202", "#BCC3CE", "#F49538", "#7D5329")
  palette_numbers <- c("#5E81AC", "#0D98BA", "#66CDAA", "#98FB98", "#E7D202", "#F49538", "#BF616A")
  palette_labels  <- c("#5E81AC", "#8FA87A", "#BF616A", "#E7D202", "#7D5329", "#F49538", "#66CDAA", "#D070B9", "#98FB98", "#FCA3B7")
  palette_missing <- "#BCC3CE"

  list(fg              = "#3b4252",
       adj             = 0.5,
       ann             = TRUE,
       bg              = "white",
       bty             = "o",
       cex             = 1,
       cex.axis        = 1.2,
       cex.lab         = 1.5,
       cex.main        = 2,
       cex.sub         = 1.5,
       col             = "#3b4252",
       col.axis        = "#3b4252",
       col.lab         = "#3b4252",
       col.main        = "#3b4252",
       col.sub         = "#3b4252",
       family          = "mono",
       font            = 2,
       font.axis       = 1,
       font.lab        = 2,
       font.main       = 2,
       font.sub        = 2,
       lab             = c(4,4,7),
       las             = 1,
       lend            = 0,
       ljoin           = 0,
       lmitre          = 10,
       lty             = 1,
       lwd             = 4,
       mgp             = c(2.5,0.25,0),
       pch             = 15,
       tck             = 0,
       xaxs            = "r",
       xaxt            = "s",
       xpd             = FALSE,
       yaxs            = "r",
       yaxt            = "s",
       rect.col        = NA,
       rect.density    = numeric(),
       rect.angle      = 45,
       rect.lwd        = 4,
       rect.lty        = 1,
       rect.border     = NA,
       palette         = palette,
       palette.numbers = palette_numbers,
       palette.labels  = palette_labels,
       palette.missing = palette_missing
       )
}


basetheme_ink <- function() {
  palette         <- c("#000000", "#FF0000", "#008B00", "#0000FF", "#00FFFF", "#EE1289", "#ECC700", "#778899", "#FF8C00", "#8B4513")
  palette_numbers <- c("#0000FF", "#20B2AA", "#00FFFF", "#00FF00", "#ECC700", "#FF8C00", "#FF0000")
  palette_labels  <- c("#FF0000", "#0000FF", "#008B00", "#8B4513", "#ECC700", "#A020F0", "#FF8C00", "#EE1289", "#00FFFF", "#00FF00")
  palette_missing <- "#778899"

  list(fg              = "blue",
       adj             = 0.5,
       ann             = TRUE,
       bg              = "navajowhite",
       bty             = "o",
       cex             = 0.8,
       cex.axis        = 1,
       cex.lab         = 1,
       cex.main        = 1.5,
       cex.sub         = 1,
       col             = "blue",
       col.axis        = "blue",
       col.lab         = "blue",
       col.main        = "blue",
       col.sub         = "blue",
       family          = "",
       font            = 3,
       font.axis       = 3,
       font.lab        = 4,
       font.main       = 2,
       font.sub        = 3,
       lab             = c(5, 5, 7),
       las             = 1,
       lend            = "round",
       ljoin           = "round",
       lmitre          = 10,
       lty             = "dotted",
       lwd             = 2,
       mgp             = c(2, 0.5, 0),
       pch             = 4,
       tck             = -0.01,
       xaxs            = "r",
       xaxt            = "s",
       xpd             = FALSE,
       yaxs            = "r",
       yaxt            = "s",
       rect.col        = "gray80",
       rect.density    = 10,
       rect.angle      = 45,
       rect.lwd        = 1,
       rect.lty        = 1,
       rect.border     = NA,
       palette         = palette,
       palette.numbers = palette_numbers,
       palette.labels  = palette_labels,
       palette.missing = palette_missing
       )
}


basetheme_dark <- function() {
  palette         <- c("#BAC1BF", "#ED2939", "#66C84D", "#3870C2", "#40E0D0", "#FC0FCD", "#F2D63B", "#777B7E", "#F2A83B", "#8A6842")
  palette_numbers <- c("#3870C2", "#00B7EB", "#40E0D0", "#66C84D", "#F2D63B", "#FC6600", "#ED2939")
  palette_labels  <- c("#3870C2", "#8AC1D4", "#F2A83B", "#66C84D", "#DA2E20", "#FFFD55", "#EB8677", "#A6EB99", "#DF89E8", "#8A6842")
  palette_missing <- c("#777B7E")

  list(fg              = "#7E848C",
       adj             = 0.5,
       ann             = TRUE,
       bg              = "#2E3947",
       bty             = "n",
       cex             = 0.8,
       cex.axis        = 1,
       cex.lab         = 1,
       cex.main        = 1.5,
       cex.sub         = 1,
       col             = "#BEBEBE",
       col.axis        = "#7E848C",
       col.lab         = "#BEBEBE",
       col.main        = "#EFF0F2",
       col.sub         = "#737D89",
       family          = "",
       font            = 1,
       font.axis       = 1,
       font.lab        = 2,
       font.main       = 2,
       font.sub        = 3,
       lab             = c(5, 5, 7),
       las             = 1,
       lend            = "round",
       ljoin           = "round",
       lmitre          = 10,
       lty             = 1,
       lwd             = 1,
       mgp             = c(3, 0.7, 0),
       pch             = 19,
       tck             = -0.01,
       xaxs            = "r",
       xaxt            = "s",
       xpd             = FALSE,
       yaxs            = "r",
       yaxt            = "s",
       rect.col        = NA,
       rect.density    = numeric(),
       rect.angle      = 45,
       rect.lwd        = 1,
       rect.lty        = 1,
       rect.border     = NA,
       palette         = palette,
       palette.numbers = palette_numbers,
       palette.labels  = palette_labels,
       palette.missing = palette_missing
       )
}

basetheme_deepblue <- function() {
  palette         <- c("#B9BBB6", "#C4512C", "#42BA90", "#3A82E4", "#00CCCC", "#EC5578", "#BEE948", "#777B7E", "#C38A13", "#795C32")
  palette_numbers <- c("#3A82E4", "#00B7EB", "#00CCCC", "#25CE7A", "#BEE948", "#C38A13", "#C4512C")
  palette_labels  <- c("#42BA90", "#BEE948", "#3A82E4", "#D25D38", "#E9A820", "#9041BA", "#795C32", "#EC5578", "#00B7EB", "#FEE12B")
  palette_missing <- "#777B7E"

  list(fg              = "#CCCCCC",
       adj             = 0.5,
       ann             = TRUE,
       bg              = "#002E49",
       bty             = "l",
       cex             = 1,
       cex.axis        = 0.8,
       cex.lab         = 0.8,
       cex.main        = 1,
       cex.sub         = 0.8,
       col             = "#BEBEBE",
       col.axis        = "#BEBEBE",
       col.lab         = "#FFFFFF",
       col.main        = "#FFFFFF",
       col.sub         = "#326D87",
       family          = "",
       font            = 1,
       font.axis       = 1,
       font.lab        = 2,
       font.main       = 2,
       font.sub        = 3,
       lab             = c(5, 5, 7),
       las             = 1,
       lend            = "round",
       ljoin           = "round",
       lmitre          = 10,
       lty             = 1,
       lwd             = 2,
       mgp             = c(3, 0.7, 0),
       pch             = 21,
       tck             = -0.015,
       xaxs            = "r",
       xaxt            = "s",
       xpd             = FALSE,
       yaxs            = "r",
       yaxt            = "s",
       rect.col        = NA,
       rect.density    = numeric(),
       rect.angle      = 45,
       rect.lwd        = 1,
       rect.lty        = 1,
       rect.border     = NA,
       palette         = palette,
       palette.numbers = palette_numbers,
       palette.labels  = palette_labels,
       palette.missing = palette_missing
       )
}

basetheme_royal <- function() {

  palette         <- c("#E5DDCE", "#B33941", "#2E8B57", "#1175BB", "#4AC6AE", "#E93FBC", "#DBA520", "#777B7E", "#EF7215", "#997950")
  palette_numbers <- c("#1175BB", "#5097A4", "#4AC6AE", "#88BE6D", "#E4CD05", "#EF7215", "#B33941")
  palette_labels  <- c("#7852A9", "#E93FBC", "#B33941", "#1175BB", "#2E8B57", "#DBA520", "#EF7215", "#4AC6AE", "#997950", "#E27069")
  palette_missing <- "#777B7E"

  palette <- c("#1175BB", "#B33941", "#EE6450", "#7D40AE", "#266D1F", "#FADA5E", "#5A4D41")

  list(fg              = "#DACEC4",
       adj             = 0.5,
       ann             = TRUE,
       bg              = "#2D2D3B",
       bty             = "o",
       cex             = 1,
       cex.axis        = 0.8,
       cex.lab         = 0.8,
       cex.main        = 1.2,
       cex.sub         = 0.8,
       col             = "#DACEC4",
       col.axis        = "#DACEC4",
       col.lab         = "#FFFFFF",
       col.main        = "#FFFFFF",
       col.sub         = "#DACEC4",
       family          = "serif",
       font            = 1,
       font.axis       = 1,
       font.lab        = 2,
       font.main       = 2,
       font.sub        = 3,
       lab             = c(5, 5, 7),
       las             = 1,
       lend            = "round",
       ljoin           = "round",
       lmitre          = 10,
       lty             = 1,
       lwd             = 1.2,
       mgp             = c(3, 0.7, 0),
       pch             = 21,
       tck             = -0.015,
       xaxs            = "r",
       xaxt            = "s",
       xpd             = FALSE,
       yaxs            = "r",
       yaxt            = "s",
       rect.col        = NA,
       rect.density    = numeric(),
       rect.angle      = 45,
       rect.lwd        = 1,
       rect.lty        = 1,
       rect.border     = NA,
       palette         = palette,
       palette.numbers = palette_numbers,
       palette.labels  = palette_labels,
       palette.missing = palette_missing
       )
}

list_themes <- function() {
  c(brutal   = basetheme_brutal,
    clean    = basetheme_clean,
    default  = basetheme_default,
    ink      = basetheme_ink,
    minimal  = basetheme_minimal,
    dark     = basetheme_dark,
    deepblue = basetheme_deepblue,
    royal    = basetheme_royal,
    void     = basetheme_void
    )
}
